[dtensor] enable foreach operators for adam optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112108

This PR enables basic foreach ops in DTensor for adam optimizer, to improve performance
compare to optimizer using torch.Tensor. Currently by default optimizer
won't do this for tensor subclass, we will need to enable this by
default in DTensor when all ops are covered, or early enable this when
exploring new FSDP, we just need to append DTensor to the optimizer
allow list.

Some latency measurement, on a 5-layer MLP model:
single tensor adam: 17ms
![Screenshot 2023-10-29 at 10 48 22 PM](https://github.com/pytorch/pytorch/assets/9443650/8937d786-b863-4318-88c2-12e43180ce8d)
foreach multitensor adam: 4ms
![Screenshot 2023-10-29 at 10 50 58 PM](https://github.com/pytorch/pytorch/assets/9443650/de105cc3-8e12-4765-938a-763d8e958194)

so around 4.25x improvement

cc @awgu

wanchaol committed:
[dtensor] enable foreach operators for adam optimizer

This PR enables basic foreach ops in DTensor for adam optimizer, to improve performance
compare to optimizer using torch.Tensor. Currently by default optimizer
won't do this for tensor subclass, we will need to enable this by
default in DTensor when all ops are covered, or early enable this when
exploring new FSDP, we just need to append DTensor to the optimizer
allow list.

cc @awgu

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] enable foreach operators for adam optimizer"


This PR enables basic foreach ops in DTensor for adam optimizer, to improve performance
compare to optimizer using torch.Tensor. Currently by default optimizer
won't do this for tensor subclass, we will need to enable this by
default in DTensor when all ops are covered, or early enable this when
exploring new FSDP, we just need to append DTensor to the optimizer
allow list.

cc awgu

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] enable foreach operators for adam optimizer"


This PR enables basic foreach ops in DTensor for adam optimizer, to improve performance
compare to optimizer using torch.Tensor. Currently by default optimizer
won't do this for tensor subclass, we will need to enable this by
default in DTensor when all ops are covered, or early enable this when
exploring new FSDP, we just need to append DTensor to the optimizer
allow list.

Some latency measurement, on a 5-layer MLP model:
single tensor adam: 17ms
![Screenshot 2023-10-29 at 10 48 22 PM](https://github.com/pytorch/pytorch/assets/9443650/8937d786-b863-4318-88c2-12e43180ce8d)
foreach multitensor adam: 4ms
![Screenshot 2023-10-29 at 10 50 58 PM](https://github.com/pytorch/pytorch/assets/9443650/de105cc3-8e12-4765-938a-763d8e958194)

so around 4.25x improvement

cc awgu

[ghstack-poisoned]

wz337 reviewed (approved):
LGTM!

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -71,14 +71,18 @@ def __str__(self) -> str:
             )
         output_spec_str = self.pretty_print_placements(self.output_spec.placements)
-        return f"({input_specs_str}) -> ({output_spec_str}) @ mesh layout: {tuple(self.output_spec.mesh.mesh.shape)}"
+        return f"{{{input_specs_str}{output_spec_str}}}"
```
curious: any reason we want to have the curly braces? i.e. why do we prefer `{(input1, input2) -> output}` over`(input1, input2) -> output`?

wanchaol replied:
I don't quite remember why lol, maybe the later is also good, I'll try to change this in follow up PRs.

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -121,26 +126,24 @@ def output_ndim(self):
     """
     TupleStrategy represents the output strategy of this op is a tuple
-    of strategy, i.e. If the output of this op is a tuple of tensors, we should
-    return a TupleStrategy that contains a tuple of OpStrategy.
+    of strategy, i.e. If the output of this op is a tuple of tensors or list of tensors
+    with possibly different placement strategies, we should return a TupleStrategy that
+    contains a tuple of OpStrategy.
 
-    NOTE: if the output of the op is a List[Tensor], it's likely we should return
-    OpStrategy directly in all cases.
+    NOTE: if the output of the op is a List[Tensor] and they share the same placement
+    strategy, then we should return a single OpStrategy instead of a TupleStrategy
```
Why do we want to minimize the size of output strategy? Any performance regression if we returned a list of identical strategies instead of one OpStrategy for a list of tensors??

wanchaol replied:
I documented this so that we won't confuse with whether to use OpStrategy or TupleStrategy. If it's a placement that every output of list have the same strategy, we should use OpStrategy for simplicity reason and avoid any perf concerns, I don't think it would regress the perf but it would be good to keep TupleStrategy to only be used in certain cases.

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -121,26 +126,24 @@ def output_ndim(self):
-                for idx, strat in enumerate(self.childs)
-            ]
+        tuple_strategies_str = "TupleStrategy("
+        child_strategies_str = ", ".join(
+            [f"{str(strat)}" for idx, strat in enumerate(self.childs)]
         )
-        return f"{tuple_strategies_str}\n{child_strategies_str}"
+        return f"{tuple_strategies_str}{child_strategies_str})"
```
maybe change to
```
child_strategies_str = ", ".join(...)
return f"TupleStrategy({child_strategies_str})"
```
for a better readability since we only use `tuple_strategies_str` once? Just my personal opinion when reading the code.

wanchaol replied:
Yeah, sure, will submit follow up PR to avoid trigger CI again

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -209,21 +212,26 @@ def __repr__(self) -> str:
             if isinstance(arg, DTensorSpec):
                 args_sharding.append(str(arg))
-                mesh = arg.mesh
+                mesh_shape = tuple(arg.mesh.mesh.shape)
```
suggestion: since we already have `device_mesh.ndim` and `device_mesh.size`, shall we add `device_mesh.shape` to avoid this `mesh.mesh.shape`?

wanchaol replied:
We can do that yeah, I'll leave it as a follow up PR

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -209,21 +212,26 @@ def __repr__(self) -> str:
                 arg_spec = arg.strategies[0].output_spec
                 args_sharding.append(str(arg_spec))
-                mesh = arg_spec.mesh
+                mesh_shape = tuple(arg_spec.mesh.mesh.shape)
```
same as above

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -209,21 +212,26 @@ def __repr__(self) -> str:
-                mesh = arg_spec.mesh
+                mesh_shape = tuple(arg_spec.mesh.mesh.shape)
+            elif isinstance(arg, TupleStrategy):
+                first_op_strtgy = arg.childs[0]
+                assert isinstance(first_op_strtgy, OpStrategy)
+                mesh_shape = tuple(
+                    first_op_strtgy.strategies[0].output_spec.mesh.mesh.shape
+                )
```
`first_op_strtgy` -> `first_op_strategy`?

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "[dtensor] enable foreach operators for adam optimizer"


This PR enables basic foreach ops in DTensor for adam optimizer, to improve performance
compare to optimizer using torch.Tensor. Currently by default optimizer
won't do this for tensor subclass, we will need to enable this by
default in DTensor when all ops are covered, or early enable this when
exploring new FSDP, we just need to append DTensor to the optimizer
allow list.

Some latency measurement, on a 5-layer MLP model:
single tensor adam: 17ms
![Screenshot 2023-10-29 at 10 48 22 PM](https://github.com/pytorch/pytorch/assets/9443650/8937d786-b863-4318-88c2-12e43180ce8d)
foreach multitensor adam: 4ms
![Screenshot 2023-10-29 at 10 50 58 PM](https://github.com/pytorch/pytorch/assets/9443650/de105cc3-8e12-4765-938a-763d8e958194)

so around 4.25x improvement

cc awgu

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

awgu commented:
@wanchaol What is the technical complexity to further enable fused Adam?

wanchaol commented:
> @wanchaol What is the technical complexity to further enable fused Adam?

@awgu It should be relatively easy, I haven't try that yet but if you need this I can take a look soon
